### PR TITLE
KAFKA-5031: validate count of records for DefaultRecordBatch in validateMessagesAndAssignOffsets

### PR DESCRIPTION
**Pull request by gosubpl**
https://issues.apache.org/jira/browse/KAFKA-5031

Implements additional check for `DefaultRecordBatch` that compares number of records declared in the header with actual number of records.

### CODE REVIEW

**Comment by gosubpl** [comment]
Added code is covered by both `LogValidatorTest` and `LogTest` for various scenarios (including compacted topics).

File: clients/src/main/java/org/apache/kafka/common/record/InvalidBatchSizeException.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.errors.CorruptRecordException;
+
+public class InvalidBatchSizeException extends CorruptRecordException {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/InvalidBatchSizeException.java]: We should probably just use `InvalidRecordException` since we are not introducing a new error code for this.

File: core/src/main/scala/kafka/log/LogValidator.scala [hunk 2]
```diff
@@ -126,9 +127,15 @@ private[kafka] object LogValidator extends Logging {
...
       validateBatch(batch, isFromClient, toMagicValue)
 
+      var recordCount = 0
       for (record <- batch.asScala) {
         validateRecord(batch, record, now, timestampType, timestampDiffMaxMs, compactedTopic)
         builder.appendWithOffset(offsetCounter.getAndIncrement(), record)
+        recordCount += 1
+      }
+      if (batch.isInstanceOf[DefaultRecordBatch]) {
+        if(batch.asInstanceOf[DefaultRecordBatch].count() != recordCount)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogValidator.scala]: Instead of doing this we can add a `countOrNull` method that returns `null` for the old message formats. It would also be good if we can share the logic a bit more.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. A few comments:

1. It would be good to have tests for these checks.
2. It seems like we are not checking the header count?

**Comment by ijuma** [comment]
@gosubpl, will you be able to address the review comments in the next couple of days? Otherwise, someone else will have to pick it up so that it can be included in the release.

**Comment by gosubpl** [comment]
Working on this as we speak :)

**Comment by gosubpl** [comment]
@ijuma I have basically only two (hopefully simple) questions:
* there is no obvious value in the arguments with which the header count should be reconciled; can you please advise what should that be? Shall the number of headers also be compared to the number of validated records?
* please confirm that I have correctly understood, that you'd prefer object oriented approach to the `countOrNull` implementation - implement it in the whole inheritance hierarchy and have return `null` wherever it does not make sense

**Comment by ijuma** [comment]
@gosubpl, looking into this more carefully, it seems like it is a bit more involved. If we iterate over `RecordBatch`, it will always succeed because the `count` is used to end the iteration. It seems like we'd need to update the batch iterator to verify that we reached the end of the buffer after reading the last message. The headers check would be similar.

cc @hachikuji in case he has a different opinion.

**Comment by gosubpl** [comment]
@ijuma @hachikuji Yes, I have just realized the same. My idea for now would be to move those sanity checks to the `DefaultRecordBatch` somewhere, to validate that everything makes sense, because if the number of records matches the expected one, and we have all headers then the rest of the checks will be done on them by the `validateMessagesAndAssignOffsets`.
WDYT?

**Comment by hachikuji** [comment]
@gosubpl Yes, that makes sense to me. When iterating the batch, we can keep track of the number of records and raise an error if we reach the end earlier than expected or if the expected count is reached and there are still remaining bytes. The header check can do something similar in `DefaultRecord`.

**Comment by gosubpl** [comment]
@hachikuji thanks for assistance & clarification. I will add assertions that validate the correctness of number of records in the `DefaultRecordBatch` to all its iterators and explicit header check for the `DefaultRecord`. Hope to submit PR today.

**Comment by gosubpl** [comment]
Please do not merge yet, still WIP. I have added proper batch size validation, need to add headers check, will do so later today.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 3]
```diff
@@ -221,8 +222,45 @@ public int partitionLeaderEpoch() {
         return buffer.getInt(PARTITION_LEADER_EPOCH_OFFSET);
     }
 
+    public boolean validateBatchSize() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Hmm... Seems rather wasteful to decompress another time just to validate the record count. What I had thought is that we could hook the logic into `DefaultRecordBatch.RecordIterator` so that we can do this validation as part of the single pass we do over the records when validating. As a matter of fact, we seem to already do part of the validation there. If the number of records that we consumed is greater than the count, then we raise `NoSuchElementException`. That should probably be changed to `InvalidRecordException` and then we just need to check for the underflow case.

Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: @hachikuji Underflow case is not difficult. But the problem with the overflow is that `hasNext` has the condition `readRecords < numRecords` so the case for overflow in `next()` which is `readRecords >= numRecords` will never be checked. What could be done is to force call on next again after going through the whole recordset (without checking `hasNext`) to see if this triggers exception... (good) or doesn't (bad). This also could be a separate function. WDYT?

**Comment by gosubpl** [comment]
@hachikuji @ijuma I have integrated batch record count checks into the `DefaultRecordBatch.RecordIterator` and added header count checks for `DefaultRecord`. Would be great if you could take a look at it.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 4]
```diff
@@ -350,27 +351,38 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
...
+                headers[i] = new RecordHeader(headerKey, headerValue);
             }
-
-            headers[i] = new RecordHeader(headerKey, headerValue);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid number of headers declared for the record");
         }
-
-        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+        if (buffer.position() != buffer.limit())
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Hmm... This check seems to depend on the buffer only containing data from this record. We should either duplicate the buffer and set the limit according to the expected size of the record, or verify that the position delta (i.e. the difference of the final and initial positions) is equal to the expected size.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 5]
```diff
@@ -350,27 +351,38 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
...
+                    headerValue.limit(headerValueSize);
+                    buffer.position(buffer.position() + headerValueSize);
+                }
+
+                headers[i] = new RecordHeader(headerKey, headerValue);
             }
-
-            headers[i] = new RecordHeader(headerKey, headerValue);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid number of headers declared for the record");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: I think perhaps the error message should be more generic. An invalid number of headers is one possibility, but it could also just be incomplete data.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 6]
```diff
@@ -500,7 +508,19 @@ public Record next() {
                 throw new NoSuchElementException();
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
+                boolean isValid = false;
+                try {
+                    readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: This is a little odd. I wonder if we can instead add an abstract `hasRemaining()` method and use `InputStream.available()` and `ByteBuffer.hasRemaining()` to determine whether there are still bytes left. cc @ijuma

**Review by hachikuji — COMMENTED** [review]
A few more comments.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 4 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Yup. Just realized that the buffer is not exclusive for a given record. Will fix.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 6 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Sure. Will try this approach.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: One issue is that `available` doesn't have the contract we want. See the following comment for more details.

https://github.com/apache/kafka/pull/2025#issuecomment-260653068

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Yeah, @ijuma is right. Unfortunately, we may have to live without this check until we have a better input interface to work with. For a little additional context, we have found that the overhead of throwing `EOFException` is excessive and we took pains to eliminate it, so it would be unfortunate if we still need it in the end.

The consequence of not having this check is that a buggy or malicious producer could write additional data at the end of the batch which would not be reachable. It pollutes the log and may be a difficult issue for a client developer to debug, but it doesn't actually cause any problems that I can think of, so maybe we can live with it.

Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Hmm... I will try approaching this via a `read()` call. Let me see if this is practicable.

**Comment by gosubpl** [comment]
@hachikuji @ijuma I have managed to fix the bug in the headers check and also hopefully sorted out the issue with checking for bytes left at the end of the batch - my approach is via additional `read()` call for the stream case, as was kindly suggested in the JDK ticket that was linked to the case you highlighted ( http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7031075 ).

**Comment by gosubpl** [comment]
Not sure how to re-trigger failed build. Can `org.apache.kafka.streams.integration.EosIntegrationTest.shouldNotViolateEosIfOneTaskFailsWithState` be a flaky tests? It passes without problems locally for JDK 1.8 + Scala 2.12

**Comment by gosubpl** [comment]
@ijuma @hachikuji ready for another round of review, AFAIK :)

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 7]
```diff
@@ -350,27 +352,39 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
...
             }
-
-            headers[i] = new RecordHeader(headerKey, headerValue);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid header data or number of headers declared for the record");
         }
-
-        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+        // validate whether we have read all header bytes in the current record
+        if (buffer.position() - recordStart != sizeInBytes - ByteUtils.sizeOfVarint(sizeInBytes))
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: nit: We could probably just pass `sizeOfBodyInBytes` into `readFrom` and this function.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 8]
```diff
@@ -245,12 +246,23 @@ public int partitionLeaderEpoch() {
...
                     return DefaultRecord.readFrom(inputStream, baseOffset, baseTimestamp, baseSequence, logAppendTime);
+                } catch (EOFException e) {
+                    throw new InvalidRecordException("Incorrect declared batch size, premature EOF reached");
                 } catch (IOException e) {
                     throw new KafkaException("Failed to decompress record stream", e);
                 }
             }
 
             @Override
+            protected boolean eofReached() {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: nit: the name is a little awkward. How about `hasRemaining` to match `ByteBuffer`?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 9]
```diff
@@ -497,14 +517,22 @@ public boolean hasNext() {
...
         public Record next() {
             if (readRecords >= numRecords)
-                throw new NoSuchElementException();
+                throw new InvalidRecordException("Garbled data resulting in attempt of reading past the end of batch");
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: nit: this comment doesn't seem too useful. Maybe emphasize that this check ensures that the declared size of the batch is correct?

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala [hunk 10]
```diff
@@ -546,6 +546,42 @@ class LogValidatorTest {
     checkOffsets(messageWithOffset, offset)
   }
 
+  @Test(expected = classOf[InvalidRecordException])
+  def testInvalidRecordCountTooManyNonCompressedV2() {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: There doesn't seem to be any particular need to put these test cases here instead of DefaultRecordBatchTest. If we move them, then we can avoid the awkwardness of exposing a public static method on a test class.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 11]
```diff
@@ -497,14 +517,22 @@ public boolean hasNext() {
         @Override
         public Record next() {
             if (readRecords >= numRecords)
-                throw new NoSuchElementException();
+                throw new InvalidRecordException("Garbled data resulting in attempt of reading past the end of batch");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Seems we can revert this change. We'll hit the InvalidRecordException below if there is additional data, so this is really more for iterator sanity.

File: clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java [hunk 12]
```diff
@@ -153,6 +154,22 @@ public void testInvalidRecordSize() {
     }
 
     @Test(expected = InvalidRecordException.class)
+    public void testInvalidRecordSizeLow() {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java]: This test case doesn't seem different than `testInvalidRecordSize`.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 13]
```diff
@@ -350,27 +352,39 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
...
-            String headerKey = Utils.utf8(buffer, headerKeySize);
-            buffer.position(buffer.position() + headerKeySize);
-
-            ByteBuffer headerValue = null;
-            int headerValueSize = ByteUtils.readVarint(buffer);
-            if (headerValueSize >= 0) {
-                headerValue = buffer.slice();
-                headerValue.limit(headerValueSize);
-                buffer.position(buffer.position() + headerValueSize);
+        try {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Seems like this buffer underflow could be hit in `readFrom` as well when reading the various fields. Maybe we can relocate the try block?

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Left a few more comments.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 13 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Sure.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 7 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: We are already passing it. `sizeInBytes` is `ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes` (passed into `readFrom`). In this check we are trying to find out whether actually `sizeOfBodyInBytes` (minus `sizeOfVarint(sizeOfBodyInBytes)` to account for this count itself) bytes have been read from the `buffer` in this record.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 8 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Great! I like it :)

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 11 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Sure.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 9 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Yes, thanks.

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala (hunk 10 above)
Comment (gosubpl) [reply core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: Yes, all moved.

File: clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java (hunk 12 above)
Comment (gosubpl) [reply clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java]: One of them is for record size too high, the other - for too low. I have clarified this by changing names appropriately.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 14]
```diff
@@ -315,41 +316,52 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
...
+            if (numHeaders == 0)
+                return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
 
-        if (numHeaders == 0)
-            return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
+            Header[] headers = getHeaders(buffer, numHeaders, recordStart, sizeInBytes);
 
+            return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid header data or number of headers declared for the record");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: The try/catch is over the whole block, so the message seems a little odd. If this can only happen while reading the headers, we should probably move the try/catch to `getHeaders`. We should also probably rename `geHeader` to `readHeaders` or something like that.

File: clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java [hunk 15]
```diff
@@ -153,6 +154,67 @@ public void testInvalidRecordSize() {
     }
 
     @Test(expected = InvalidRecordException.class)
+    public void testInvalidRecordSizeTooLow() {
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java]: Both of these cases are still testing the same thing. I think you are intending to set an invalid record count, but this is actually changing the size of the batch (i.e. the size in bytes). So whether it is 2 or 10, we're validating the same path.

Comment (gosubpl) [reply clients/src/test/java/org/apache/kafka/common/record/DefaultRecordBatchTest.java]: Thanks for spotting! Before refactoring that test was covering different case. Now the cases of invalid record count are being covered by the relocated `testInvalidRecordCount*` tests.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 16]
```diff
@@ -245,12 +247,23 @@ public int partitionLeaderEpoch() {
...
+                    throw new InvalidRecordException("Incorrect declared batch size, premature EOF reached");
                 } catch (IOException e) {
                     throw new KafkaException("Failed to decompress record stream", e);
                 }
             }
 
             @Override
+            protected boolean hasRemaining() {
+                try {
+                    return inputStream.read() == -1;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Shouldn't this be inverted?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java [hunk 17]
```diff
@@ -349,7 +351,16 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
 
         if (numHeaders == 0)
             return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
+        try {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: The buffer underflow could happen above as well. Perhaps we can expand the `try` to cover this full method and make the exception more generic?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 18]
```diff
@@ -500,11 +521,21 @@ public Record next() {
...
+                // (overflow case, i.e. reading past buffer end is checked elsewhere).
+                if (!hasRemaining())
+                    throw new InvalidRecordException("Incorrect declared batch size, records still remaining in file");
+            }
+            return rec;
         }
 
         protected abstract Record readNext(long baseOffset, long baseTimestamp, int baseSequence, Long logAppendTime);
 
+        protected abstract boolean hasRemaining();
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Thinking about this name a little more... Since the InputStream version uses `read()`, this method cannot be used as generally as its name suggests. We should try to convey the fact that this check should only be used when the presumed end has been reached. Maybe something like `ensureNoneRemaining`?

**Review by hachikuji — COMMENTED** [review]
Looking good overall. Left a couple minor comments.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 16 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: The name is wrong. This should be `hasNoneRemaining()` or, even better (your suggestion) `ensureNoneRemaining()`. Previously it was called `eofReached()` but @ijuma reqested it to be changed to `hasRemaining()`, and I didn't check the condition, unfortunately :) I intend to fix by changing the name to what you suggested in the comment below. WDYT?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 18 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Yes, that is a better name, will change it.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 17 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: I have already changed this to cover the full method, but @ijuma requested it to be narrowed again to only this part. You are probably right, that more generic exception, conveying also the failures in the wider function will allow the try to span the whole function. Let me know after you both guys agree on the span of the try block.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 16 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: @hachikuji asked you to change the name originally:

```text
hachikuji 5 days ago Contributor
nit: the name is a little awkward. How about hasRemaining to match ByteBuffer?
```

:)

Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: @ijuma right, right, sorry about that. But it was you who suggested narrowing the try block. Care to comment below?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 17 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Sorry, my comment could have been clearer. I was saying that if the exception message was correct, the try block should be narrower. If the exception can be thrown from other places, it's OK for the try block to cover the whole method if we: (1) Update the message, (2) Include the original exception as the cause so that we can figure out where it was thrown from.

Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: Thanks for clarifying. Sometimes it is hard to get the exact meaning - especially that it is my first Kafka contribution. Will adjust the PR accordingly :)

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: My point is that any of the calls to `readVarint` or even `buffer.position()` prior to reading the headers could also underflow. It seems like it would be useful to raise `InvalidRecordException` consistently in all cases (not just for headers).

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 16 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Sorry for the confusion. I thought `hasRemaining` made sense initially, but then I realized that the name should be more suggestive of its usage. I'd prefer something like `ensureNoneRemaining`, but it's not a dealbreaker for me.

**Comment by gosubpl** [comment]
@ijuma @hachikuji please take a look if I got your intentions right. `hasRemaining()` has been renamed to `ensureNoneRemaining()` and I have widened the scope of `try` block in `readFrom` adding the propagation of `e.getMessage()` of original exception to the message of resultant `InvalidRecordException`.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5392/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5377/
Test PASSed (JDK 8 and Scala 2.12).

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java (hunk 17 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java]: I have adjusted that now. The range of `try` is expanded, but the detailed reason is copied to the `InvalidRecordException` message.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java (hunk 16 above)
Comment (gosubpl) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: I have changed it to `ensureNoneRemaining`. Thanks for suggestion, I think it is a good name, best of all tried up until now :)

**Review by hachikuji — APPROVED** [review]
Thanks for the patch. LGTM

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java [hunk 19]
```diff
@@ -245,12 +247,23 @@ public int partitionLeaderEpoch() {
...
                     throw new KafkaException("Failed to decompress record stream", e);
                 }
             }
 
             @Override
+            protected boolean ensureNoneRemaining() {
+                try {
+                    return inputStream.read() == -1;
+                } catch (IOException e) {
+                    return false;
```
Comment (xvrl) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java:262]: Shouldn't we be concerned about swallowing IOException here?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java:262]: Thanks for the comment. I'll address this in a follow-up.

**Comment by hachikuji** [comment]
Merging to trunk and 0.11.0.
